### ISSUE TITLE
Feature request: Dynamic override from database

### ISSUE THREAD

**Issue by tiaod** [simov/grant#110]
Passing the secret as querystring is not safe, I think maybe it's better to load the configurations from database.

```js
// First, create a provider
let provider = new Provider({
  provider: 'twitter',
  override: 'myapp',
  config: {
    "key": "...",
    "secret": "...",
    "callback": "/hi"
  }
})
provider.save()

// Then
app.use('/grant', grant({
  "defaults": {
    "protocol": "http",
    "host": "localhost:3000",
    "transport": "session",
    "state": true
  },
  "google": {
    "key": "...",
    "secret": "...",
    "scope": ["openid"],
    "nonce": true,
    "custom_params": {"access_type": "offline"},
    "callback": "/hello"
  },
  "twitter": {
    "key": "...",
    "secret": "...",
    "callback": "/hi"
  }
}, async function(provider, override){
   // Load from database.
  let doc = await Provider.findOne({
    provider,
    override
  })
  return doc.config
}))
```
- `/connect/twitter/myapp` will load the config from database.

**Comment by simov** [comment]
That's an interesting idea, I'll have to think about it.

In the meantime you can probably set your dynamic configuration directly on the `query` object that's attached to the `request` argument:


```js
var express = require('express')
var session = require('express-session')
var grant = require('grant-express')


express()
  .use(session({secret: 'grant', saveUninitialized: true, resave: true}))
  .use('/connect/:provider/:override', (req, res, next) => {
    if (
      req.params.provider === 'twitter' &&
      req.params.override === 'foo'
    ) {
      // load from database
      req.query.key = 'KEY'
      req.query.secret = 'SECRET'
    }
    next()
  })
  .use(grant(require('./config.json')))
  .get('/hello', (req, res) => {
    res.end(JSON.stringify(req.query, null, 2))
  })
  .listen(3000)
```
```json
{
  "defaults": {
    "protocol": "http",
    "host": "localhost:3000"
  },
  "twitter": {
    "callback": "/hello",
    "dynamic": true
  }
}
```

Then if you navigate to `http://localhost:3000/connect/twitter/foo` you'll be able to login using the credentials set in the middleware before Grant.

I have tested it only with Express, but it should work with Koa too.

**Comment by tiaod** [comment]
This is fine when using express 4.x. 

However, since express 5.0, they make `req.query` a getter instead of added by middleware.
https://github.com/expressjs/express/issues/3472
https://github.com/expressjs/express/issues/2215

**Comment by simov** [comment]
Yes, I was thinking about something similar. Well, it's a perfectly valid solution for JavaScript, but nowadays it looks more like a monkeypatch.

I'll definitely think about other possible solutions.

**Comment by simov** [comment]
I'm thinking about allowing the users to set the `req.session.grant` key before entering the Grant routes. That's the session key [used internally](https://github.com/simov/grant#session) by Grant.

**Comment by millette** [comment]
I also noticed that using `{ dynamic: true }` was also leaking the secret in the session (atob magic!) unless external storage was used.

I'm working on a very generic script, every provider should be configurable thru admin with no server restarts. For now, I configure grant to set all possible providers to an empty object.

```js
const config = {
  defaults: { ... },
  github: {},
  twitter: {},
  // actually, I use <https://github.com/simov/grant/blob/master/config/oauth.json> to get all provider keys
}

const grant = new Grant(config)
```

Then, `grant.config.github` is preconfigured, without a key or secret but I can set that up from a database at runtime.

It was my workaround, but I'm looking forward to a better solution :-)

P.S.: just starting with grant - thanks for the simplicity!

**Comment by simov** [comment]
@millette leaking data inside the session entirely depends on the session store implementation. It can be an in-memory one, but implemented properly with storing only the session ID inside the cookie.

You approach for lazy loading the app credentials is good, however, note that you don't need to set `dynamic: true` in this case.

`dynamic: true` is used to override config during the authorization, `grant.config` can be used to update the config for the entire instance/process, meaning it affects all authorizations anyway.

**Comment by millette** [comment]
Yes, just to clarify, first stab had a tiny defaults config with `dynamic: true`. With the pre-filled approach, I also removed `dynamic: true`.

**Comment by simov** [comment]
Dynamic input state overrides landed in v4.7.0

The documentation and examples about it are a bit scarce at the moment, but here is the relevant commit in the [readme](https://github.com/simov/grant/commit/3b04eb69a278165ae9be7ba7a06e8b85da21c5e5).

In summary now you can use the request/response lifecycle state of your HTTP framework of choice to set dynamic overrides before entering the Grant routes. This dynamic configuration works outside of the `dynamic` configuration key, meaning you can override anything using the request/response lifecycle state. The `dynamic` option controls only allowed keys to override over HTTP via GET/POST request.
